core-common/src/main/java/org/zalando/nakadi/domain/kpi/KPIEvent.java

```diff
@@ -0,0 +1,65 @@
+package org.zalando.nakadi.domain.kpi;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.generic.GenericRecordBuilder;
+import org.json.JSONObject;
+import org.zalando.nakadi.exceptions.runtime.NakadiBaseException;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.stream.Stream;
+
+public abstract class KPIEvent {
+
+    @Retention(RetentionPolicy.RUNTIME)
+    public @interface KPIField {
+        String value();
+    }
+
+    public static class KPIEventMappingException extends NakadiBaseException {
+        public KPIEventMappingException(final String message, final Throwable cause) {
+            super(message, cause);
+        }
+    }
+
+    /*
+        Implementation of this method should return the event-type name of the specific KPI event.
+        This method should not be confused with the getter "getEventType()" in some specific KPI events;
+        which is why this method is named "eventTypeOfThisKPIEvent()
+     */
+    public abstract String eventTypeOfThisKPIEvent();
+
+    private Stream<Field> getKPIFields() {
+        return Arrays.stream(this.getClass().getDeclaredFields())
+                .filter(field -> field.isAnnotationPresent(KPIField.class));
+    }
+
+    public JSONObject asJsonObject() throws KPIEventMappingException {
+        final var jsonObject = new JSONObject();
+        getKPIFields().forEach(field -> {
+            try {
+                jsonObject.put(field.getAnnotation(KPIField.class).value(), field.get(this));
```

Comment:
Nice approach with annotations! looks very clean in this way. 

Can optimize it even further - cache fields annotations for each class (usually it is done by creating/initializing mappers once per startup) and avoid using clazz.getFields() for each publishing call (there are thousands per second publishes).

Another thing - maybe we can avoid injecting mappers logic into objects (looks like mixing it alltogether, plus inheritance actually removes the beauty of annotations usage, maybe it is also possible to improve).

Reply:
Good point, I'll do this.

Reply:
Done